src/java.base/share/classes/jdk/internal/misc/Unsafe.java

```diff
@@ -3222,13 +3274,21 @@ public final Object getAndSetReference(Object o, long offset, Object newValue) {
         return v;
     }
 
-    @SuppressWarnings("unchecked")
-    public final <V> Object getAndSetFlatValue(Object o, long offset, int layout, Class<?> valueType, V newValue) {
-        synchronized (valueLock) {
-            Object oldValue = getFlatValue(o, offset, layout, valueType);
-            putFlatValue(o, offset, layout, valueType, newValue);
-            return oldValue;
-        }
+    @ForceInline
+    public final Object getAndSetReference(Object o, long offset, Class<?> valueType, Object newValue) {
+        Object v;
+        do {
+            v = getReferenceVolatile(o, offset);
+        } while (!compareAndSetReference(o, offset, valueType, v, newValue));
+        return v;
+    }
+
+    public Object getAndSetFlatValue(Object o, long offset, int layoutKind, Class<?> valueType, Object newValue) {
```

Comment:
Why no more `@ForceInline`?

Reply:
Good catch

Reply:
Still missing forceinline here
```suggestion
    @ForceInline
    public Object getAndSetFlatValue(Object o, long offset, int layoutKind, Class<?> valueType, Object newValue) {
```